### PR TITLE
mon: check for mismatched daemon versions

### PR DESCRIPTION
**Pull request by dzafman**
Take over with a rebase of #35876

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mon/Monitor.cc
Comment (batrick) [inline src/mon/Monitor.cc:5342]:
```suggestion
  for (auto& [rank, metadata] : mon_metadata) {
    auto& pv = metadata.at("ceph_version");
    auto& ph = metadata.at("hostname");
    versions[pv.second][i] = std::make_pair(rank, ph.second);
```

would be easier to read.

File: src/mon/HealthMonitor.h
Comment (batrick) [inline src/mon/HealthMonitor.h:25]: Better would be `ceph::coarse_mono_time`.

File: src/mon/HealthMonitor.cc
Comment (batrick) [inline src/mon/HealthMonitor.cc:694]: This loop never runs?

**Review by batrick — DISMISSED** [review]

File: src/mon/Monitor.cc
Comment (dzafman) [reply src/mon/Monitor.cc]: @bpatrick If you like to look at the latest code you can see if your suggestion may apply partially.

**Comment by dzafman** [comment]
My output looks like this:

```
HEALTH_ERR There is a daemon running an older version of ceph

[WRN] DAE_INCORRECT_VERSION: There is a daemon running an older version of ceph
    osd.2 is running an older version of ceph: ceph version 15.0.0.1111-g48f95b9d9ac 
```

When multiple versions and multiple daemons the wording changes.

```
HEALTH_ERR There are daemons running old versions of ceph
[WRN] DAE_INCORRECT_VERSION: There are daemons running old versions of ceph
    osd.0 is running an older version of ceph: 14.0.0-4573-g58f95b9d9ac
mon.a is running an older version of ceph: 15.0.0-4573-g58f95b9d9ac
```

**Comment by dzafman** [comment]
My latest output looks like this:

```
HEALTH_WARN There is a daemon running an older version of ceph
[WRN] DAE_OLDER_VERSION: There is a daemon running an older version of ceph
    osd.1 is running an older version of ceph: 01.00.00-gversion-test
```

For multiple versions gives an HEALTH_ERR instead of warning and add "multiple" in the message.

```
HEALTH_ERR There are daemons running multiple old versions of ceph
[ERR] DAE_OLDER_VERSION: There are daemons running multiple old versions of ceph
    osd.1 osd.2 are running an older version of ceph: 01.00.00-gversion-test
    osd.0 is running an older version of ceph: 02.00.00-gversion-test
```

**Comment by dzafman** [comment]
@tchaikov When you are available, could you look at what I need to do for Crimson that is causing compilation errors in this pull requests.

```
[ 96%] Building CXX object src/crimson/osd/CMakeFiles/crimson-osd.dir/osd_operations/peering_event.cc.o
../../../lib/libcrimson-alienstore.a(admin_socket.cc.o): In function `VersionHook::call(std::basic_string_view<char, std::char_traits<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, boost::variant<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, double, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<long, std::allocator<long> >, std::vector<double, std::allocator<double> > >, std::less<void>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, boost::variant<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, double, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<long, std::allocator<long> >, std::vector<double, std::allocator<double> > > > > > const&, ceph::Formatter*, std::ostream&, ceph::buffer::v15_2_0::list&)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/admin_socket.cc:616: undefined reference to `ceph_version_to_str(ceph::common::CephContext*)'
collect2: error: ld returned 1 exit status
src/test/crimson/CMakeFiles/unittest_seastar_alienstore_thread_pool.dir/build.make:146: recipe for target 'bin/unittest_seastar_alienstore_thread_pool' failed
make[3]: *** [bin/unittest_seastar_alienstore_thread_pool] Error 1
CMakeFiles/Makefile2:28801: recipe for target 'src/test/crimson/CMakeFiles/unittest_seastar_alienstore_thread_pool.dir/all' failed
make[2]: *** [src/test/crimson/CMakeFiles/unittest_seastar_alienstore_thread_pool.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
```

File: src/common/version.h
Comment (tchaikov) [inline src/common/version.h:22]: this function is also used by crimson. you might want to update `src/crimson/admin/admin_socket.cc` to pass `nullptr` to `ceph_version_to_str()`.

**Comment by dzafman** [comment]
As a follow on we need to figure out why the mgr isn't sending the version modified by the test configuration variable.

File: src/mon/Monitor.cc
Comment (dzafman) [reply src/mon/Monitor.cc]: I made the change in the new location where similar code exists.

**Comment by dzafman** [comment]
@tchaikov Could you are someone in the crimson team fix the make check problem?  After a rebase a new crimson compilation error has occurred.

**Comment by tchaikov** [comment]
@dzafman will take a look.

**Comment by jdurgin** [comment]
can we spell out DAEMON_OLD_VERSION instead of abbreviating, and add it to the docs about health checks?

**Comment by dzafman** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
@dzafman 

```diff
diff --git a/src/common/admin_socket.cc b/src/common/admin_socket.cc
index 23ef34ce13c..aa642fa243e 100644
--- a/src/common/admin_socket.cc
+++ b/src/common/admin_socket.cc
@@ -626,7 +626,9 @@ public:
     } else {
       f->open_object_section("version");
       if (command == "version") {
+#ifndef WITH_ALIEN
    f->dump_string("version", ceph_version_to_str(nullptr));
+#endif
    f->dump_string("release", ceph_release_to_str());
    f->dump_string("release_type", ceph_release_type());
       } else if (command == "git_version") {
```

because both alien store and crimson use this `ceph_version_to_str()`, so we need to link alienstore against `version.o` compiled from `version.cc`. but they actually expect different versions of this function

- alien store expects the classic one, where `CephContext` is a full-blown version. 
- crimson expects the crimson variant, where `CephContext` is but a dummy/minimal implementation.

when we were facing this scenario back when we worked on porting alien store to client, a namespace is added to differentiate these shared symbols. but i'd just disable this output in alienstore this time, as alienstore does not handle asock command at all at this moment.

**Comment by neha-ojha** [comment]
@dzafman is this just waiting on https://github.com/ceph/ceph/pull/36759#issuecomment-704132563?

**Comment by dzafman** [comment]
@neha-ojha I've included the fix from Kefu.  It should be ready to merge if Github checks pass.

**Comment by neha-ojha** [comment]
> @neha-ojha I've included the fix from Kefu. It should be ready to merge if Github checks pass.

has it already gone through teuthology? we should see how this does with the upgrade tests and also use them to verify this feature.

File: doc/rados/operations/health-checks.rst
Comment (jdurgin) [inline doc/rados/operations/health-checks.rst:32]: we should be able to avoid this during an upgrade, e.g. only warn after a certain period of time. users shouldn't need to take any action in the majority of cases

Comment (dzafman) [reply doc/rados/operations/health-checks.rst:32]: I think we talked about having the upgrade recipes to do something to mark upgrade in progress and clear when upgrade complete.  However, for a manual update my doc would be appropriate.  Can we leave this to a follow on fix when the different recipes can be changed?

Comment (jdurgin) [reply doc/rados/operations/health-checks.rst:32]: yes that could be addressed in a follow-up, just want to make sure we fix it up before pacific is released

**Comment by dzafman** [comment]
@jdurgin If we don't at least add Ansible upgrade handing we get test errors because of wait_for_healthy use when DAEMON_OLDER_VERSION health warning is occurring.  Does teuthology always use Ansible?  At least for upgrade testing?

Example:
http://qa-proxy.ceph.com/teuthology/dzafman-2020-10-23_22:00:18-upgrade-wip-zafman-46264-2-distro-basic-smithi/5552287/teuthology.log

**Comment by neha-ojha** [comment]
> @jdurgin If we don't at least add Ansible upgrade handing we get test errors because of wait_for_healthy use when DAEMON_OLDER_VERSION health warning is occurring. Does teuthology always use Ansible? At least for upgrade testing?
> 
> Example:
> http://qa-proxy.ceph.com/teuthology/dzafman-2020-10-23_22:00:18-upgrade-wip-zafman-46264-2-distro-basic-smithi/5552287/teuthology.log

We can introduce an option called `mon_warn_on_mismatched_versions` and set it to false for upgrade tests. We can expand it further to only warn after a threshold amount of time.

**Comment by dzafman** [comment]
@neha-ojha I already have "mon_warn_on_older_version" config value we can set to false.

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [reply doc/rados/operations/health-checks.rst:32]: @dzafman I'd expect this warning to catch intentional or unintentional incomplete upgrades, by muting the warning during upgrades, we essentially lose its purpose. I'd like this warning to trigger after a reasonable amount of time, we can go with a safe default.

Comment (dzafman) [reply doc/rados/operations/health-checks.rst:32]: I'm not convinced there is a safe default for the trigger delay that would work because of the variation in cluster sizes.

In order for teuthology upgrade tests to work I'm making cephadm mute DAE_OLD_VERSION warnings at the start of upgrade and unmute after upgrade completion.  If you pause upgrade, I'm thinking of leaving the warning muted, but I'm open to change that.  We could unmute on pause and mute on resume.

The documentation text above would be changed to "Unless you use cephadm to upgrade, this will be an expected warning while in the middle of an upgrade."  The plan is to the the mute/unmute other upgrade handlers like Ansible and update the doc accordingly.

File: PendingReleaseNotes
Comment (jdurgin) [inline PendingReleaseNotes:25]:
```suggestion
  upgrading has finished use "ceph health unmute DAEMON_OLD_VERSION".
```

File: doc/rados/operations/health-checks.rst
Comment (jdurgin) [inline doc/rados/operations/health-checks.rst:38]:
```suggestion
upgrading has finished use "ceph health unmute DAEMON_OLD_VERSION".
```

**Review by jdurgin — APPROVED** [review]
looks good. we may want to store the timestamp on disk in the future, but in memory is much better than today

**Comment by dzafman** [comment]
Testing PASSED:

https://pulpito.ceph.com/dzafman-2020-11-09_17:18:57-upgrade-wip-zafman-testing-46264-2-distro-basic-smithi
Rerun 5 dead jobs
https://pulpito.ceph.com/dzafman-2020-11-10_09:38:54-upgrade-wip-zafman-testing-46264-2-distro-basic-smithi/

**Comment by dzafman** [comment]
jenkins test make check

**Comment by dzafman** [comment]
jenkins test make check

**Comment by dzafman** [comment]
jenkins test make check

File: PendingReleaseNotes
Comment (neha-ojha) [inline PendingReleaseNotes:20]:
```suggestion
  This condition must exist for over mon_warn_older_version_delay (set to 1 week by default) in order for the
```

File: PendingReleaseNotes
Comment (neha-ojha) [inline PendingReleaseNotes:22]:
```suggestion
  without falsely seeing the warning.  If upgrade is paused for an extended
```

File: PendingReleaseNotes
Comment (neha-ojha) [inline PendingReleaseNotes:25]:
```suggestion
  upgrade has finished use "ceph health unmute DAEMON_OLD_VERSION".
```

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst:33]:
```suggestion
This condition must exist for over mon_warn_older_version_delay (set to 1 week by default) in order for the
```

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst:35]:
```suggestion
without falsely seeing the warning.  If upgrade is paused for an extended
```

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst:38]:
```suggestion
upgrade has finished use "ceph health unmute DAEMON_OLD_VERSION".
```

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:48]: This comment seems incorrect now.

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:97]: ditto

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:152]: ditto

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:172]: why two separate sleeps?

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:40]: nit: could add a description above each TEST_check_version_health_* to explain what it is meant to do

File: qa/standalone/misc/ver-health.sh
Comment (neha-ojha) [inline qa/standalone/misc/ver-health.sh:56]: nit: do we need mds for this test?

File: PendingReleaseNotes
Comment (neha-ojha) [inline PendingReleaseNotes:19]:
```suggestion
  on daemons. It will generate a health error if multiple versions are detected.
```

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst:31]:
```suggestion
Warn if old version(s) of Ceph are running on any daemons.
```

File: doc/rados/operations/health-checks.rst
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst:32]:
```suggestion
It will generate a health error if multiple versions are detected.
```

File: PendingReleaseNotes
Comment (neha-ojha) [inline PendingReleaseNotes:18]:
```suggestion
* A new health check, DAEMON_OLD_VERSION, will warn if different versions of Ceph are running
```

**Review by neha-ojha — COMMENTED** [review]
few minor comments and 97996474c479f44cb70243b1ffb94abd6ed8dc68 seems to a merge commit?

File: qa/standalone/misc/ver-health.sh
Comment (dzafman) [reply qa/standalone/misc/ver-health.sh:172]: I wanted to keep the sleep for kill to finish and the 50% of the 20 second delay separate.

**Review by neha-ojha — APPROVED** [review]
new changes look good, needs signed-off and squashing to remove merge commit

**Comment by neha-ojha** [comment]
jenkins test api

**Comment by neha-ojha** [comment]
@dzafman this PR is causing failures like https://pulpito.ceph.com/teuthology-2020-11-12_07:01:02-rados-master-distro-basic-smithi/5615313/

**Comment by tobias-urdin** [comment]
I think this healthcheck has a bug that it treats 16.2.10 as older than 16.2.9
Reported here https://tracker.ceph.com/issues/58166

**Pull request merged**
